turbopack/xtask/src/patch_package_json.rs

```diff
@@ -0,0 +1,235 @@
+    patch_workspace_package_json_map(args, package_json_map)?;
+    write_json_value(&package_json_path, &package_json_value)?;
+
+    println!("Successfully patched {package_json_path:?} to use local tarball files!");
+    // TODO: Bun is currently broken for unknown reasons, it generates ENOTDIR when trying to read
+    // the tarballs
+    // NOTE: Yarn is currently broken due to an upstream issue:
+    // https://github.com/yarnpkg/yarn/issues/6339
+    println!("Run `pnpm i` or `npm i` to install the overrides");
+
+    Ok(())
+}
+
+fn read_json_value(path: &Path) -> Result<Value> {
+    serde_json::from_reader(BufReader::new(
+        File::open(path).with_context(|| format!("could not read {path:?}"))?,
+    ))
+    .with_context(|| format!("failed to parse {path:?}"))
+}
+
+fn write_json_value(path: &Path, value: &Value) -> Result<()> {
+    (|| {
+        let file = File::options().write(true).truncate(true).open(path)?;
+        let mut writer = BufWriter::new(file);
+
+        // NOTE: serde_json defaults to two-space indents
+        serde_json::to_writer_pretty(&mut writer, value)?;
+        writer.write_all(b"\n")?;
+        writer.flush()?;
+        anyhow::Ok(())
+    })()
+    .with_context(|| format!("failed to write {path:?}"))
+}
+
+fn to_package_json_map(value: &mut Value) -> Result<&mut JsonMap> {
+    value
+        .as_object_mut()
+        .context("package.json must represent an object")
+}
+
+fn patch_workspace_package_json_map(
+    args: &PatchPackageJsonArgs,
+    package_json_map: &mut JsonMap,
+) -> Result<()> {
+    let next_peer_deps = get_next_peer_deps()?;
+
+    // insert overrides
+    let overrides = [
+        ("next", &*format!("file:{}", args.next_tarball)),
+        ("@next/mdx", &*format!("file:{}", args.next_mdx_tarball)),
+        ("@next/env", &*format!("file:{}", args.next_env_tarball)),
+        (
+            "@next/bundle-analyzer",
+            &*format!("file:{}", args.next_bundle_analyzer_tarball),
+        ),
+        // next-swc is added to the project's package.json, but also set a global override just in
+        // case something else pulls it in
+        ("@next/swc", &*format!("file:{}", args.next_swc_tarball)),
+        // next's peerDependencies
+        ("react", &*next_peer_deps.react),
+        ("react-dom", &*next_peer_deps.react_dom),
+    ];
+
+    // npm uses `overrides`: https://docs.npmjs.com/cli/v10/configuring-npm/package-json#overrides
+    // bun also supports this: https://bun.sh/docs/install/overrides
+    let overrides_map = get_mut_or_insert_default_object(package_json_map, "overrides")?;
+    insert_map_entries(overrides_map, &overrides[..]);
+
+    // pnpm uses `pnpm.overrides`: https://pnpm.io/package_json#pnpmoverrides
+    let pnpm_map = get_mut_or_insert_default_object(package_json_map, "pnpm")?;
+    let pnpm_overrides_map = get_mut_or_insert_default_object(pnpm_map, "overrides")?;
+    insert_map_entries(pnpm_overrides_map, &overrides[..]);
```

Comment:
FYI pnpm supports `resolutions` and `pnpm.overrides` (we used `resolutions` for the longest time in Next.js until https://github.com/vercel/next.js/pull/68186/files). Not sure what happens if you have both.